src/yubikey.rs

```diff
@@ -839,36 +826,41 @@ impl YubiKey {
                     "won't set new key '{:?}' since it's weak (with odd parity)",
                     new_key
                 );
-                res = Err(Error::KeyError);
-            } else {
-                apdu.ins = YKPIV_INS_SET_MGMKEY;
-                apdu.p1 = 0xff;
-
-                apdu.p2 = match touch {
-                    0 => 0xff,
-                    1 => 0xfe,
-                    _ => {
-                        let _ = self._ykpiv_end_transaction();
-                        return Err(Error::GenericError);
-                    }
-                };
 
-                apdu.lc = DES_LEN_3DES as u8 + 3;
-                apdu.data[0] = YKPIV_ALGO_3DES;
-                apdu.data[1] = YKPIV_KEY_CARDMGM;
-                apdu.data[2] = DES_LEN_3DES as u8;
-                apdu.data[3..3 + DES_LEN_3DES].copy_from_slice(new_key);
-
-                res = self._send_data(&mut apdu, data.as_mut_ptr(), &mut recv_len, &mut sw);
+                let _ = self._ykpiv_end_transaction();
+                return Err(Error::KeyError);
+            }
 
-                if res.is_ok() && sw != SW_SUCCESS {
-                    res = Err(Error::GenericError);
+            let p2 = match touch {
+                0 => 0xff,
+                1 => 0xfe,
+                _ => {
+                    let _ = self._ykpiv_end_transaction();
+                    return Err(Error::GenericError);
                 }
+            };
+
+            let mut data = [0u8; DES_LEN_3DES + 3];
+            data[0] = YKPIV_ALGO_3DES;
+            data[1] = YKPIV_KEY_CARDMGM;
+            data[2] = DES_LEN_3DES as u8;
+            data[3..3 + DES_LEN_3DES].copy_from_slice(new_key);
+
+            let apdu = APDU::new(YKPIV_INS_SET_MGMKEY)
+                .params(0xff, p2)
+                .data(&data)
+                .to_bytes();
+
+            let mut data = [0u8; 261];
+            let mut recv_len = data.len() as u32;
+            let mut sw: i32 = 0;
+            res = self._send_data(apdu.as_slice(), data.as_mut_ptr(), &mut recv_len, &mut sw);
+
+            if res.is_ok() && sw != SW_SUCCESS {
+                res = Err(Error::GenericError);
             }
         }
 
-        apdu.zeroize();
-        let _ = self._ykpiv_end_transaction();
```

Comment:
Now missing the transaction end for the paths that lead here.

Reply:
@str4d wouldn't worry too much about these. in my PC/SC PR I'm deleting them all and replacing them with the safe RAII `pcsc::Transaction` type.